csrc/fusion.cpp

```diff
@@ -26,21 +26,24 @@
 
 namespace nvfuser {
 
-static thread_local Fusion* ACTIVE_FUSION = nullptr; // NOLINT
+/*static*/ thread_local Fusion* FusionGuard::active_fusion_ = nullptr;
 
-FusionGuard::FusionGuard(Fusion* fusion) : prev_fusion{ACTIVE_FUSION} {
-  ACTIVE_FUSION = fusion;
+FusionGuard::FusionGuard(const Fusion* fusion) : prev_fusion_(active_fusion_) {
```

Comment:
Does this has to be constant pointer?

Reply:
This was added to accommodate https://github.com/NVIDIA/Fuser/blob/75c2cd9ed4c3bce693df07bf2f3646bf39a8d08a/csrc/fusion.cpp#L506C24-L506C24. I could `const_cast` over there. But `FusionGuard` taking a const pointer shows that `FusionGuard` *itself* has no interests modifying the input fusion. Users of `getCurFusion` can still modify it but that's a separate concern.

Reply:
Could we just do:

```C++
void Fusion::printTransforms() {
  FUSER_PERF_SCOPE("Fusion::printTransforms");
  IrTransformPrinter t_exprs(debug());
  t_exprs.handle(this);
}
```
there? `IrTransformPrinter` will eventually call into https://github.com/NVIDIA/Fuser/blob/75c2cd9ed4c3bce693df07bf2f3646bf39a8d08a/csrc/iter_visitor.cpp#L140 which has a fusion guard already. `IrTransformPrinter` also has a `handle(const Fusion*)` which just const-cast away the const and call `handle(Fusion*)`.

Reply:
You are essentially recommending to remove the FusionGuard in `Fusion::printTransforms`. You may be right, but I don't know how to prove this or even test this. We want to make sure `FusionGuard::getCurFusion` is always called with a FusionGuard set up. It looks like all of our unit tests that build fusions start with a FusionGuard. The python API also sets up `FusionGuard` [here](https://github.com/NVIDIA/Fuser/blob/75c2cd9ed4c3bce693df07bf2f3646bf39a8d08a/csrc/python_frontend/fusion_state.cpp#L73). Does that mean we don't even need the FusionGuard in `IterVisitor::traverseBetween`?

Reply:
> Does that mean we don't even need the FusionGuard in `IterVisitor::traverseBetween`?

I don't think we have a well defined guidelines on when should we use `FusionGuard`, when not. It is just an ad-hoc approach that when people see something failing because of no IR container, people just add a `FusionGuard` wherever convenient. For PRs, generally as long as CI is green, it is fine.

On the other hand, I disagree with this statement:

> But `FusionGuard` taking a const pointer shows that `FusionGuard` _itself_ has no interests modifying the input fusion.

This is just like `std::shared_ptr` itself is not interested in modifying the managed data, but this does not mean that `std::shared_ptr<T>`'s ctor should take a `const T*` instead of `T*`, especially it will have to const cast it away.

Reply:
Fixed -- I reverted my FusionGuard-taking-const-pointers changes.